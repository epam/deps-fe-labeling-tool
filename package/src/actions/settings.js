import { createAction } from 'redux-actions'

const FEATURE_NAME = 'SETTINGS'

const storeSettings = createAction(`${FEATURE_NAME}/STORE`)

const resetDefault = createAction(`${FEATURE_NAME}/RESET_DEFAULT`)

export {
  resetDefault,
  storeSettings
}
