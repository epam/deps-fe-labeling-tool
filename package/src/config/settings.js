import { Settings } from '@/models/Settings'

const settings = new Settings()

export {
  settings
}
