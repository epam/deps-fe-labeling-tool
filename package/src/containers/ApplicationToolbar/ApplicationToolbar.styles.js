import styled, { css } from 'styled-components'
import { COLORS } from '@/theme/theme.default'

const separatorPseudo = css`
  content: ' ';
  height: 3rem;
  width: 1px;
  position: absolute;
  background: ${COLORS.SHADOW_1};
  top: 0.7rem;
`

const Wrapper = styled.div`
  display: grid;
  grid-template-columns: 6fr 4fr 4fr;
  padding: 0.7rem;

  .ant-btn {
    margin: 0 2px;
  }
`

const Block = styled.div`
  display: flex;
  align-items: center;
`

const RightBlock = styled.div`
  display: flex;
  justify-content: flex-end;
  align-items: center;
`

const RightSeparator = styled.div`
  line-height: 0;

  &:not(:first-child) {
    margin-left: 1.3rem;
  }

  &::after {
    ${separatorPseudo}
    margin: 0.5rem 0.6rem;
  }
`

const LeftSeparator = styled.div`
  line-height: 0;

  &:not(:last-child) {
    margin-right: 1.3rem;
  }

  &::before {
    ${separatorPseudo}
    margin: 0.5rem -0.7rem;
  }
`

const Row = styled.div`
  display: inline-flex;
`

export {
  Row,
  Wrapper,
  Block,
  RightBlock,
  RightSeparator,
  LeftSeparator
}
