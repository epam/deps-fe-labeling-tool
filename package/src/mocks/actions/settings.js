const storeSettings = jest.fn((settings) => ({
  type: 'MOCK_STORE_SETTINGS_ACTION',
  payload: settings
}))

const mockSettingsActions = {
  storeSettings
}

export {
  mockSettingsActions
}
