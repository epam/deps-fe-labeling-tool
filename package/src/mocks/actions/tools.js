const changeTool = jest.fn((tool) => ({
  type: 'MOCK_CHANGE_TOOL_ACTION',
  payload: tool
}))

const mockToolsActions = {
  changeTool
}

export {
  mockToolsActions
}
