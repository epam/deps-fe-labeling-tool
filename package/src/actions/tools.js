import { createAction } from 'redux-actions'
import { clearSelection } from '@/actions/markup'
import { setActiveSidebar } from '@/actions/ui'
import { SidebarContent } from '@/enums/SidebarContent'
import { Settings } from '@/models/Settings'
import { currentPageSelector } from '@/selectors/pagination'
import { settingsSelector } from '@/selectors/settings'
import { selectedToolSelector } from '@/selectors/tools'

const FEATURE_NAME = 'TOOLS'

const selectTool = createAction(
  `${FEATURE_NAME}/SELECT_TOOL`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const changeTool = (toolToSelect) => (dispatch, getState) => {
  const state = getState()
  const settings = settingsSelector(state)
  if (!Settings.has(settings, toolToSelect)) {
    return
  }

  const currentPage = currentPageSelector(state)
  const selectedTool = selectedToolSelector(state)
  if (selectedTool !== toolToSelect) {
    dispatch(setActiveSidebar(SidebarContent.MARKUP))
    dispatch(clearSelection(currentPage))
    dispatch(selectTool(toolToSelect))
  }
}

export {
  resetDefault,
  selectTool,
  changeTool
}
