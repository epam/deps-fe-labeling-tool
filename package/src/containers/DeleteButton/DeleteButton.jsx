import React, { useCallback, useEffect } from 'react'
import { ExclamationCircleOutlined } from '@ant-design/icons'
import { Modal } from 'antd'
import 'antd/lib/modal/style/index.less'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { removeLabels, removeTables } from '@/actions/markup'
import { setActiveSidebar } from '@/actions/ui'
import { Button } from '@/components/Button'
import { DeleteIcon } from '@/components/Icons/DeleteIcon'
import { Tooltip } from '@/components/Tooltip'
import { HotKeyEvent } from '@/constants/hotKeys'
import { SidebarContent } from '@/enums/SidebarContent'
import { withHotKeys } from '@/hocs/withHotKeys'
import { labelShape } from '@/models/Label'
import { tableShape } from '@/models/Table'
import {
  pageSelectedLabelsSelector,
  pageSelectedTablesSelector,
  pageLabelsSelector
} from '@/selectors/markup'
import { currentPageSelector } from '@/selectors/pagination'
import { replaceAll } from '@/utils/string'

const RESOURCE_TOOLTIP_DELETE = 'Delete'
const RESOURCE_CONFIRM_DELETE_TITLE = 'Deletion Confirmation'
const RESOURCE_CONFIRM_DELETE_CONTENT = 'Are you sure you want to delete markup?'
const RESOURCE_CONFIRM_OBJECTS_DELETION_CONTENT_1 = 'Are you sure you want to delete {0} objects?'
const RESOURCE_CANCEL_DELETE = 'Cancel'

const DeleteButton = ({
  currentPage,
  selectedLabels,
  selectedTables,
  removeLabels,
  removeTables,
  registerHandlers,
  setActiveSidebar
}) => {
  const deleteMarkup = useCallback(() => {
    selectedLabels.length && removeLabels(currentPage, selectedLabels)
    selectedTables.length && removeTables(currentPage, selectedTables)
  }, [
    removeLabels,
    removeTables,
    selectedLabels,
    selectedTables,
    currentPage
  ])

  const confirmDelete = useCallback((onOk, onCancel, content) => Modal.confirm({
    title: RESOURCE_CONFIRM_DELETE_TITLE,
    icon: <ExclamationCircleOutlined />,
    content: content,
    okText: RESOURCE_TOOLTIP_DELETE,
    cancelText: RESOURCE_CANCEL_DELETE,
    onOk,
    onCancel
  }), [])

  const onClick = useCallback((e, confirmContent = RESOURCE_CONFIRM_DELETE_CONTENT) => {
    confirmDelete(
      deleteMarkup,
      () => { },
      confirmContent
    )
  }, [confirmDelete, deleteMarkup])

  const onKeyDelete = useCallback(() => {
    const selectedObjects = [
      ...selectedLabels,
      ...selectedTables
    ]

    if (selectedObjects.length === 0) {
      return
    }

    if (selectedObjects.length === 1) {
      setActiveSidebar(SidebarContent.MARKUP)
      deleteMarkup()
      return
    }

    onClick(null, replaceAll(RESOURCE_CONFIRM_OBJECTS_DELETION_CONTENT_1, selectedObjects.length))
  }, [
    selectedLabels,
    selectedTables,
    deleteMarkup,
    onClick,
    setActiveSidebar
  ])

  const isDisabled = () => !(
    selectedLabels.length ||
    selectedTables.length
  )

  useEffect(() => {
    const deleteHotKeyHandler = {
      [HotKeyEvent.DELETE]: () => onKeyDelete()
    }
    registerHandlers(deleteHotKeyHandler)
  }, [
    onKeyDelete,
    registerHandlers
  ])

  return (
    <Tooltip title={RESOURCE_TOOLTIP_DELETE}>
      <Button.Icon
        disabled={isDisabled()}
        icon={<DeleteIcon />}
        onClick={onClick}
      />
    </Tooltip>
  )
}

DeleteButton.propTypes = {
  currentPage: PropTypes.number.isRequired,
  selectedLabels: PropTypes.arrayOf(labelShape).isRequired,
  selectedTables: PropTypes.arrayOf(tableShape).isRequired,
  removeLabels: PropTypes.func.isRequired,
  removeTables: PropTypes.func.isRequired,
  registerHandlers: PropTypes.func.isRequired,
  setActiveSidebar: PropTypes.func.isRequired
}

const mapDispatchToProps = {
  removeLabels,
  removeTables,
  setActiveSidebar
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state),
  labels: pageLabelsSelector(state),
  selectedLabels: pageSelectedLabelsSelector(state),
  selectedTables: pageSelectedTablesSelector(state)
})

const ConnectedComponent = withHotKeys(
  connect(mapStateToProps, mapDispatchToProps)(DeleteButton)
)

export {
  ConnectedComponent as DeleteButton
}
