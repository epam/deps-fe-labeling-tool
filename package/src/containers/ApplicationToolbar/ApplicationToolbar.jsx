
import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { CompilationFeatureControl } from '@/components/CompilationFeatureControl'
import { CloseButton } from '@/containers/CloseButton'
import { CommentsButton } from '@/containers/CommentsButton'
import { DeleteButton } from '@/containers/DeleteButton'
import { DocumentPageSwitcher } from '@/containers/DocumentPageSwitcher'
import { DocumentProperties } from '@/containers/DocumentProperties'
import { HotKeysDrawer } from '@/containers/HotKeysDrawer'
import { ImageActions } from '@/containers/ImageActions'
import { MainMenu } from '@/containers/MainMenu'
import { ObjectActions, isObjectActionsEmpty } from '@/containers/ObjectActions'
import { QuickAccessActions } from '@/containers/QuickAccessActions'
import { ToolsPicker, isToolsPickerEmpty } from '@/containers/ToolsPicker'
import { CompilationFeature } from '@/enums/CompilationFeature'
import { Feature } from '@/enums/Feature'
import { Tool } from '@/enums/Tool'
import { Settings, settingsShape } from '@/models/Settings'
import { settingsSelector } from '@/selectors/settings'
import { selectedToolSelector } from '@/selectors/tools'
import {
  Wrapper,
  Row,
  Block,
  RightBlock,
  LeftSeparator,
  RightSeparator
} from './ApplicationToolbar.styles'

const ApplicationToolbar = ({
  settings,
  selectedTool
}) => {
  const [hotKeysVisible, setHotKeysVisible] = useState(false)

  const toggleHotKeysInfo = () => {
    setHotKeysVisible((prevVisible) => !prevVisible)
  }

  return (
    <Wrapper>
      <Block>
        <RightSeparator>
          <MainMenu
            toggleHotKeysInfo={toggleHotKeysInfo}
          />
        </RightSeparator>
        <RightSeparator>
          <QuickAccessActions />
        </RightSeparator>
        {
          !isToolsPickerEmpty(settings) && (
            <RightSeparator>
              <Row>
                <ToolsPicker />
              </Row>
            </RightSeparator>
          )
        }
        {
          selectedTool === Tool.POINTER && (
            <RightSeparator>
              <DeleteButton />
            </RightSeparator>
          )
        }
        {
          !isObjectActionsEmpty(settings) && (
            <RightSeparator>
              <ObjectActions />
            </RightSeparator>
          )
        }
      </Block>
      <Block>
        <ImageActions />
      </Block>
      <RightBlock>
        {
          Settings.has(settings, Feature.PAGING) && (
            <LeftSeparator>
              <DocumentPageSwitcher />
            </LeftSeparator>
          )
        }
        <LeftSeparator>
          <DocumentProperties />
        </LeftSeparator>
        <CompilationFeatureControl featureName={CompilationFeature.SHOW_NOT_IMPLEMENTED}>
          <LeftSeparator>
            <CommentsButton />
          </LeftSeparator>
        </CompilationFeatureControl>
        <LeftSeparator>
          <CloseButton />
        </LeftSeparator>
      </RightBlock>
      <HotKeysDrawer
        visible={hotKeysVisible}
        toggleHotKeysInfo={toggleHotKeysInfo}
      />
    </Wrapper>
  )
}

ApplicationToolbar.propTypes = {
  settings: settingsShape,
  selectedTool: PropTypes.oneOf(
    Object.values(Tool)
  ).isRequired
}

const mapStateToProps = (state) => ({
  settings: settingsSelector(state),
  selectedTool: selectedToolSelector(state)
})

const Container = connect(mapStateToProps)(ApplicationToolbar)

export {
  Container as ApplicationToolbar
}
