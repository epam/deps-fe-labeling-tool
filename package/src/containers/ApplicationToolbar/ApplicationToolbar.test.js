import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockSettingsSelectors } from '@/mocks/selectors/settings'
import { mockToolsSelectors } from '@/mocks/selectors/tools'
import React from 'react'
import { shallow } from 'enzyme'
import { ApplicationToolbar } from '@/containers/ApplicationToolbar'
import { DeleteButton } from '@/containers/DeleteButton'
import { Tool } from '@/enums/Tool'

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/selectors/settings', () => mockSettingsSelectors)
jest.mock('@/selectors/tools', () => mockToolsSelectors)

jest.mock('@/containers/MainMenu', () => ({
  ...mockComponent('MainMenu'),
  isMainMenuEmpty: jest.fn(() => false)
}))

jest.mock('@/containers/ObjectActions', () => ({
  ...mockComponent('ObjectActions'),
  isObjectActionsEmpty: jest.fn(() => false)
}))

jest.mock('@/containers/ToolsPicker', () => ({
  ...mockComponent('ToolsPicker'),
  isToolsPickerEmpty: jest.fn(() => false)
}))

jest.mock('@/containers/HotKeysDrawer', () => mockComponent('HotKeysDrawer'))
jest.mock('@/containers/DeleteButton', () => mockComponent('DeleteButton'))
jest.mock('@/containers/ImageActions', () => mockComponent('ImageActions'))
jest.mock('@/containers/DocumentPageSwitcher', () => mockComponent('DocumentPageSwitcher'))
jest.mock('@/containers/DocumentProperties', () => mockComponent('DocumentProperties'))
jest.mock('@/containers/CommentsButton', () => mockComponent('CommentsButton'))
jest.mock('@/containers/CloseButton', () => mockComponent('CloseButton'))
jest.mock('@/containers/QuickAccessActions', () => mockComponent('QuickAccessActions'))

const { mapStateToProps, WrappedComponent } = ApplicationToolbar

describe('Container: ApplicationToolbar', () => {
  describe('mapStateToProps', () => {
    let props

    beforeEach(() => {
      props = mapStateToProps().props
    })

    it('should pass correct props to the wrapped component', () => {
      expect(props).toMatchSnapshot()
    })
  })

  describe('ConnectedComponent', () => {
    let defaultProps, wrapper

    beforeEach(() => {
      defaultProps = {
        settings: mockSettingsSelectors.settingsSelector(),
        selectedTool: Tool.POINTER
      }

      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render correct layout based on the props', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not render DeleteButton in case selectedTool is not equal to "pointer"', () => {
      defaultProps.selectedTool = Tool.GRABBING
      wrapper.setProps(defaultProps)
      expect(wrapper.find(DeleteButton).exists()).toBe(false)
    })
  })
})
