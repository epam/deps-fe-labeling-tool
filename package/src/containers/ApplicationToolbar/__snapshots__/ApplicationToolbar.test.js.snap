// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: ApplicationToolbar ConnectedComponent should render correct layout based on the props 1`] = `
<ApplicationToolbarstyles__Wrapper>
  <ApplicationToolbarstyles__Block>
    <ApplicationToolbarstyles__RightSeparator>
      <MainMenu
        toggleHotKeysInfo={[Function]}
      />
    </ApplicationToolbarstyles__RightSeparator>
    <ApplicationToolbarstyles__RightSeparator>
      <QuickAccessActions />
    </ApplicationToolbarstyles__RightSeparator>
    <ApplicationToolbarstyles__RightSeparator>
      <ApplicationToolbarstyles__Row>
        <ToolsPicker />
      </ApplicationToolbarstyles__Row>
    </ApplicationToolbarstyles__RightSeparator>
    <ApplicationToolbarstyles__RightSeparator>
      <DeleteButton />
    </ApplicationToolbarstyles__RightSeparator>
    <ApplicationToolbarstyles__RightSeparator>
      <ObjectActions />
    </ApplicationToolbarstyles__RightSeparator>
  </ApplicationToolbarstyles__Block>
  <ApplicationToolbarstyles__Block>
    <ImageActions />
  </ApplicationToolbarstyles__Block>
  <ApplicationToolbarstyles__RightBlock>
    <ApplicationToolbarstyles__LeftSeparator>
      <DocumentPageSwitcher />
    </ApplicationToolbarstyles__LeftSeparator>
    <ApplicationToolbarstyles__LeftSeparator>
      <DocumentProperties />
    </ApplicationToolbarstyles__LeftSeparator>
    <CompilationFeatureControl
      featureName="SHOW_NOT_IMPLEMENTED"
    >
      <ApplicationToolbarstyles__LeftSeparator>
        <CommentsButton />
      </ApplicationToolbarstyles__LeftSeparator>
    </CompilationFeatureControl>
    <ApplicationToolbarstyles__LeftSeparator>
      <CloseButton />
    </ApplicationToolbarstyles__LeftSeparator>
  </ApplicationToolbarstyles__RightBlock>
  <HotKeysDrawer
    toggleHotKeysInfo={[Function]}
    visible={false}
  />
</ApplicationToolbarstyles__Wrapper>
`;

exports[`Container: ApplicationToolbar mapStateToProps should pass correct props to the wrapped component 1`] = `
Object {
  "selectedTool": "label",
  "settings": Settings {
    "features": Array [
      SubSettings {
        "code": "autoSave",
        "data": Object {
          "interval": 180000,
        },
      },
      "export",
      "import",
      SubSettings {
        "code": "paging",
        "data": Object {
          "initialPage": 1,
        },
      },
      "language",
      "rotation",
      SubSettings {
        "code": "specialSymbols",
        "data": Array [
          "&",
          "$",
          "£",
          "¥",
          "©",
          "®",
          "¢",
          "%",
          "*",
          "@",
          "µ",
          "±",
          "´",
          "²",
          "~",
          "½",
          "«",
          "»",
          "°",
          "–",
          "—",
          "“",
          "”",
        ],
      },
      "manageFields",
      "multiAssignLabels",
      "multiAssignTables",
    ],
    "mode": "default",
    "panels": Array [
      "TOOLBAR",
      "MARKUP_SIDEBAR",
      "LEFT_SIDEBAR",
    ],
    "tools": Array [
      "detectTables",
      "label",
      "table",
      "merge",
      "split",
      "pointer",
      "grabbing",
    ],
  },
}
`;
